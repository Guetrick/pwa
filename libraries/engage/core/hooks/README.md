# Shopgate ENGAGE React Hooks

This library provides you with a collecton of usefull React Hooks that you can use to build out new features in Shopgate ENGAGE.

## Content

### `useCustomRoute`

This Hook returns all necessary React elements that you need to build a custom route:

* `View` - The view component that you should wrap around your view content
* `Appbar` - The app bar component for the appropriate theme (GMD or iOS),
* `params` - The current Route's parameters.

#### Usage

```jsx
import { useCustomRoute, Route } from '@shopgate/engage/core';

function MyCustomView() {
  const { View, AppBar, params } = useCustomRoute();

  return (
    <View>
      <AppBar title="My Custom View" />
      {/* ... your custom content goes here */}
    </View>
  )
}

function MyCustomRoute() {
  return (
    <Route pattern="/my-custom-url" component={MyCustomView} />
  );
}

export default MyCustomRoute;
```
