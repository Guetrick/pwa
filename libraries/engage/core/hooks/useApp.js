import { useContext } from 'react';
import AppContext from '../contexts/AppContext';

/**
 * Provides the properties of the app context.
 * @returns {Object}
 */
export function useApp() {
  const appContext = useContext(AppContext);
  return appContext;
}
