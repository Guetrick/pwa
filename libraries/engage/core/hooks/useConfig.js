import appConfig from '@shopgate/pwa-common/helpers/config';

/**
 * Provides the global theme configuration.
 * @returns {Object}
 */
export function useConfig() {
  const { theme: { pages, ...config } = {} } = appConfig;
  return config || {};
}
