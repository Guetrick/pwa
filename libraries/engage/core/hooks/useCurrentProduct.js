import { useContext } from 'react';
import { ThemeContext } from '@shopgate/pwa-common/context';

/**
 * Provides the current product context props.
 * @returns {Object}
 */
export function useCurrentProduct() {
  // The contexts are left out in favor of other hooks.
  const { contexts: { ProductContext } } = useContext(ThemeContext);
  const productProps = useContext(ProductContext);
  return productProps;
}
