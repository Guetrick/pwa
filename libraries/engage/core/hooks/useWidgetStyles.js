import { useWidgetConfig } from './useWidgetConfig';

/**
 * Retrieves the widget-configuration-styles for a specific widget by its id and an optional index.
 * If no index is given it will return the first found widget-configuration-styles by widgetId.
 * @param {string} widgetId The ID of the widget to look for.
 * @param {number|undefined} [index] The optional index of the widget.
 * @returns {Object}
 */
export function useWidgetStyles(widgetId, index) {
  const { styles = {} } = useWidgetConfig(widgetId, index);
  return styles;
}
