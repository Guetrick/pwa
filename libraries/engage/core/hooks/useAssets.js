import { useConfig } from './useConfig';

/**
 * Retrieves the global theme assets.
 * @returns {Object}
 */
export function useAssets() {
  const { assets = {} } = useConfig();

  return assets;
}
