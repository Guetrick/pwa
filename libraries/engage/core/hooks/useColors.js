import { useConfig } from './useConfig';

/**
 * Retrieves the global theme colors.
 * @returns {Object}
 */
export function useColors() {
  const { colors = {} } = useConfig();

  return colors;
}
