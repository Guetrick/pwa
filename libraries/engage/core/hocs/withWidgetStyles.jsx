import React from 'react';
import { useWidgetStyles } from '../hooks/useWidgetStyles';

/**
 * Retrieves the styles for a specific widget by its ID.
 * @param {Function} WrappedComponent The react component to wrap.
 * @param {string} widgetId The id of the widget to look for, which must exist in the config.
 * @param {number|undefined} [index] The optional index of the widget.
 * @returns {Object}
 */
export function withWidgetStyles(WrappedComponent, widgetId, index) {
  /**
   * The actual HOC.
   * @param {Object} props The component props.
   * @returns {JSX}
   */
  const WithWidgetStyles = (props) => {
    const styles = useWidgetStyles(widgetId, index);
    return (<WrappedComponent widgetStyles={styles} {...props} />);
  };

  const displayName = WrappedComponent.displayName || WrappedComponent.name || 'Component';
  WithWidgetStyles.displayName = `WithWidgetStyles(${displayName})`;

  return WithWidgetStyles;
}
