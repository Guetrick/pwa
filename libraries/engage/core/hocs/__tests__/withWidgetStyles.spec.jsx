import React from 'react';
import { mount } from 'enzyme';
import { withWidgetStyles } from '../withWidgetStyles';
import { useWidgetStyles } from '../../hooks/useWidgetStyles';

jest.mock('../../hooks/useWidgetStyles', () => ({
  useWidgetStyles: jest.fn(),
}));

// eslint-disable-next-line require-jsdoc
const MockComponent = () => null;

const widgetStyles = {
  some: 'style',
};

describe('engage > core > hocs > withWidgetStyles', () => {
  beforeEach(() => {
    useWidgetStyles.mockReturnValue(widgetStyles);
  });

  it('should merge add the widget settings to a wrapped component', () => {
    const ComposedComponent = withWidgetStyles(MockComponent);
    const wrapper = mount(<ComposedComponent />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).prop('widgetStyles')).toEqual(widgetStyles);
  });
});
