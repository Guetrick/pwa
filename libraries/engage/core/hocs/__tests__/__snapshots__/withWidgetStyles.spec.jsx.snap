// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engage > core > hocs > withWidgetStyles should merge add the widget settings to a wrapped component 1`] = `
<WithWidgetStyles(MockComponent)>
  <MockComponent
    widgetStyles={
      Object {
        "some": "style",
      }
    }
  />
</WithWidgetStyles(MockComponent)>
`;
